src/data/polynomial/mirror.lean

```diff
@@ -183,4 +184,124 @@ end
 
 end mirror
 
+section norm2
+
+variables {R : Type*} [comm_ring R] (p q : polynomial R)
+
+noncomputable def norm2 : quadratic_form R (polynomial R) :=
+finsupp.norm_sq
+
+lemma norm2_apply : p.norm2 = p.sum (λ k c, c ^ 2) :=
+finsupp.norm_sq_apply p
+
+lemma norm2_eq_sum_of_support {s : finset ℕ}
+  (h : p.support ⊆ s) : p.norm2 = s.sum (λ k, (p.coeff k) ^ 2) :=
+p.norm2_apply.trans (finset.sum_subset h (λ k h1 h2, by
+{ rw [finsupp.not_mem_support_iff.mp h2], exact zero_pow zero_lt_two }))
+
+lemma norm2_monomial (k : ℕ) (a : R) : (monomial k a).norm2 = a ^ 2 :=
+by rw [norm2_eq_sum_of_support _ (support_monomial' k a),
+  finset.sum_singleton, coeff_monomial, if_pos rfl]
+
+lemma norm2_C (a : R) : (C a).norm2 = a ^ 2 := norm2_monomial 0 a
+
+lemma norm2_zero : (0 : polynomial R).norm2 = 0 := by rw [←C_0, norm2_C, pow_two, zero_mul]
```

Comment:
Can you eliminate this and use `quadratic_form.zero_apply` instead?  I think this would actually let you elimintate `norm2_C` and `norm2_monomial`, although maybe the latter is still worth having in its finsupp version.